File: Models/patient.go
package Models

import (
	"fmt"

	"gorm.io/gorm"
)

type Patient struct {
	gorm.Model
	Name            string   `json:"name"`
	DoctorID        uint     `json:"doctor_id"`
	Address         string   `json:"address"`
	Phone           string   `json:"phone"`
	Gender          string   `json:"gender"`
	Age             int      `json:"age"`
	IsFavourite     bool     `json:"is_favourite"`
	PatientTeethMap TeethMap `json:"patient_teeth_map"`
	// Braces          Braces   `json:"braces"`
}

type Braces struct {
	gorm.Model
	PatientID          uint          `json:"patient_id"`
	IsOn               bool          `json:"is_on"`
	BracesAppointments []Appointment `json:"appointments"`
	Notes              string        `json:"notes"`
}

type TeethMap struct {
	gorm.Model
	PatientID uint
	Teeth     []Tooth `json:"teeth" gorm:"foreignKey:TeethMapID"`
}

type Tooth struct {
	gorm.Model
	ToothCode               string `json:"tooth_code"`
	TeethMapID              uint
	Condition               string        `json:"condition"`
	ConditionID             uint          `json:"condition_id"`
	IsTreated               bool          `json:"is_treated"`
	ToothHistory            []Appointment `json:"tooth_history"`
	UncompletedAppointments []Appointment `json:"uncompleted_appointments" gorm:"-"`
	HexColor                string        `json:"hex_color"`
}

func CreatePatientTeethMap(patient Patient) TeethMap {
	var teethMap = patient.PatientTeethMap
	for i := 1; i <= 8; i++ {
		teethMap.Teeth = append(teethMap.Teeth, Tooth{gorm.Model{}, fmt.Sprintf("LB%v", i), teethMap.ID, "None", 0, false, nil, nil, ""})
	}
	for i := 1; i <= 8; i++ {
		teethMap.Teeth = append(teethMap.Teeth, Tooth{gorm.Model{}, fmt.Sprintf("LT%v", i), teethMap.ID, "None", 0, false, nil, nil, ""})
	}
	for i := 1; i <= 8; i++ {
		teethMap.Teeth = append(teethMap.Teeth, Tooth{gorm.Model{}, fmt.Sprintf("RB%v", i), teethMap.ID, "None", 0, false, nil, nil, ""})
	}
	for i := 1; i <= 8; i++ {
		teethMap.Teeth = append(teethMap.Teeth, Tooth{gorm.Model{}, fmt.Sprintf("RT%v", i), teethMap.ID, "None", 0, false, nil, nil, ""})
	}
	return teethMap
}

File: lib/models/patient.dart
import 'package:dentex/main.dart';
import 'package:dentex/models/appointment.dart';
import 'package:flutter/material.dart';

class Patient {
  int id = 0;
  int doctorID = 0;
  String name = "";
  String address = "";
  String phone = "";
  String gender = "";
  int age = 0;
  bool isFavourite = false;
  Braces braces = Braces();
  TeethMap teethMap = TeethMap();
  List<Appointment> history = [];
}

class Braces {
  int id = 0;
  int teethMapID = 0;
  int patientID = 0;
  List<Appointment> appointments = [];
  String notes = "";
}

class Condition {
  int? id = 0;
  String? name = "None";
  double? price = 0.0;
  Color? color;
  Condition({this.id, this.name, this.price, this.color});
}

class Tooth {
  int id = 0;
  String toothCode = "";
  Condition condition = Condition();
  bool isTreated = false;
  List<Appointment> uncompletedAppointments = [];
  Map<String, dynamic> toJSON() => {
        "id": id,
        "tooth_code": toothCode,
        "condition": condition.name,
        "condition_id": condition.id,
        "is_treated": isTreated,
        "hex_color": HexColor(condition.color!).toHex(),
      };
}

class TeethMap {
  List<Tooth> teeth = [];
  List<Map<String, dynamic>> toJSON() {
    List<Map<String, dynamic>> jsonOutput = [];
    teeth.sort((a, b) => a.toothCode.compareTo(b.toothCode));
    for (var tooth in teeth) {
      // print(tooth.toothCode + ":" + tooth.condition.name!);
      jsonOutput.add(tooth.toJSON());
    }
    return jsonOutput;
  }
}
